compiler/src/dotty/tools/dotc/parsing/Scanners.scala

```diff
@@ -1080,16 +1080,24 @@ object Scanners {
 // Lookahead ---------------------------------------------------------------
 
     /** The next token after this one.
+     *
      *  The token is computed via fetchToken, so complex two word
      *  tokens such as CASECLASS are not recognized.
      *  Newlines and indent/unindent tokens are skipped.
      *
+     *  Since interpolation prefetches both STRINGPART and an expression,
+     *  use a scanner for rare case. Otherwise the next is overwritten on reset.
      */
-     def lookahead: TokenData =
-      if next.token == EMPTY then
+    def lookahead: TokenData =
+      if next.token != EMPTY then next
```

Comment:
I think it would be better not to burden lookahead with this logic. Can't be create a Lookahead scanner instead at the call site of `lookahead` when the assertion is triggered?

Reply:
Yes, I went back and forth on where to place the burden. There was no good answer. But the superseding PR, if it works, is conceptually much simpler, as it means lookahead never incurs the advanced token.

Reply:
I almost went down the path of, `in.canLookAhead`, because maybe there are other such states. How could I ever trust lookahead again?